### ISSUE TITLE
NVLink 통신 예제 및 필요 라이브러리

### ISSUE THREAD

**Issue by joons86** [poweraikorea/mldl-powerai#6]
안녕하세요,

NVLink를 통신을 이용하여 GPU-GPU 통신을 하려고 합니다.

간단한 예제가 있으면 따라해볼 수 있을 것 같습니다.
그리고 필요한 라이브러리, 컴파일 옵션이 있으면 알려주십시오.

감사합니다.

**Comment by nasica88** [comment]
가장 쉬운 sample code는 이미 가지고 계실 GPU 장비의 /usr/local/cuda/samples 밑에 이미 들어 있습니다.

아마 찾고 계신 sample은 아래 /usr/local/cuda/samples/0_Simple/simpleP2P 디렉토리의 것을 사용하시면 될 것입니다.

Compiler나 library는 특별히 더 필요하신 것은 없고, CUDA에 포함된 nvcc 및 cuda libraries, 그리고 gcc 정도만 있으면 됩니다.

NVLink를 위한 별도의 code는 필요없으며, GPU들이 NVLink로 연결되어 있으면 NVLink로, PCIe로 연결되어 있으면 또 PCIe 버스로 P2P 통신을 합니다.

u0017649@sys-97541:/usr/local/cuda/samples/0_Simple/simpleP2P$ pwd
/usr/local/cuda/samples/0_Simple/simpleP2P

u0017649@sys-97541:/usr/local/cuda/samples/0_Simple/simpleP2P$ ls
Makefile  NsightEclipse.xml  readme.txt  simpleP2P.cu

**Comment by joons86** [comment]
넵, 알려주신 simpleP2P 예제를 설치해서 실행해보았더니 다음과 같이 출력이 되는 걸 확인했습니다.
아래 결과에서 한가지 여쭤볼 게 있습니다..

그림에서 Checking GPU(s) for support of peer to peer memory access... 아래 각 GPU간의 연결 관계를 보여주는 곳을 보면,
GPU0 <--> GPU1
GPU2 <--> GPU3
만 메모리 엑세스가 가능하다고 나와있는데, 그럼 Minsky 머신은 GPU1 <--> GPU2 연결은 불가능한건가요?ㅠㅠ

![image](https://user-images.githubusercontent.com/43427886/55396849-d092cd00-557f-11e9-8187-d65216e68ed4.png)

**Comment by nasica88** [comment]
@joons86  예, 그 부분은 NVIDIA의 NCCL library를 이용하면 GPU1 <--> GPU2 등 모든 부분을 p2p로 연결하실 수 있습니다.  NCCL library는 장비의 CPU-GPU topolgy에 따라 최적의 경로로 (그것이 NVLink이건 CPU 소켓 간의 SMP link이건) p2p 통신을 가능하게 해줍니다.    

NCCL에 대해서는 [여기](https://developer.nvidia.com/nccl)를 참조하시기 바랍니다.

**Comment by joons86** [comment]
위 링크를 통해 NCCL 관련 문서들을 봤습니다.
가이드를 보니  커뮤니케이션 종류가

AllReduce
Broadcast
Reduce
AllGather
ReduceScatter

인 듯합니다. 현재 코드에서는 MPI_Send, MPI_Recev 같은 기능처럼 임의의 두 GPU사이에서만 통신을 하고 있는데 NCCL은 이름처럼 콜렉티브 커뮤니케이션만 지원하는 것 같습니다.ㅠ 제가 못 찾은건가요?ㅠ

저는 알려주신 p2p 샘플코드처럼 cudaMemcpy를 이용해 두 디바이스간 메모리 카피만 할 수 있으면 되는데, NCCL에서는 혹시 어떻게 이걸 구현할 수 있는지 궁금합니다.

**Comment by nasica88** [comment]
@joons86  예, NCCL을 이용하지 않고 기본적인 cudaMemcpy만 이용하는 code로는 같은 socket에 속한 2개의 GPU끼리만 통신이 가능합니다.   NCCL을 이용하지 않던 초기 caffe 버전에서는 GPU0 <-> GPU1끼리만 p2p가 되었으나, NCCL을 설치하고 LD_LIBRARY_PATH에 넣어주자 전체 4개의 GPU에서 모두 p2p가 되는 것으로 display 되더군요.

저도 개발자는 아니다보니 NCCL을 이용한 code 구현까지는 도와드리지 못하는 점 양해 부탁드립니다.
